Fix malformed packet retransmission of coap_client/coap_server

rbtchc opened the pull request:
1. A malformed packet is sent during the retransmission for both coap_client/coap_server. This is because the "appdatalen" is not set and ID + UDP headers are not dropped when retransmit. Fix the issue by doing it correctly.
1. Call net_pkt_ref() and net_pkt_unref() explicitly at retransmit_request() to avoid packet being freed right after net_context_sendto() is called.

jukkar reviewed (changes requested):
The changes look good. I would prefer that the patches are changing one logical thing at a time. So better would be if first patch fixes the appdatalen stuff both in coap client and server, and the other one the ref count in server and client.

rbtchc committed:
net: samples: fix malformed payload when coap_client/server retransmit

Original coap_client implementation does not setup "appdatalen" of
net_pkt correctly and does not strip the IP + UDP headers when doing
the retransmit. This will result in malformed coap packet. Fix it by
adding a strip_headers() function to set appdatalen and get rid of
IP + UDP headers.

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc committed:
net: samples: fix early unref issue for coap_client/server

Explicitly call net_pkt_ref()/net_pkt_unref() to avoid packet being
freed after calling net_context_sendto() at retransmit_request().
Also, do not return when net_context_sendto() returns error. Instead,
we should keep retrying.

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc commented:
Re-arrange patchset according to @jukkar 's comment.

rveerama1 commented:
LGTM.
